### ISSUE TITLE
Prevent running an upgrade rule on already-upgraded rules.

### ISSUE THREAD

**Issue by Phrohdoh** [OpenRA/OpenRA#8798]
I've planned this for quite a few months but finally decided to make an issue instead of doing it myself when I have free time (which is almost never).

My idea is a comment at the top of the file that indicates the last version that the rules were updated to.
Then the upgrader reads this line and won't run any rules that would have already been ran.
This should of course be overrideable (either via flag or just removing the line).

``` yaml
#20150614

engineer:
        Inherits: ^Infantry
        Buildable:
                Queue: Infantry
                BuildPaletteOrder: 60
        Valued:
                Cost: 500
        Tooltip:
                Name: Engineer
                Description: Captures enemy structures.\n  Unarmed
        Selectable:
                Bounds: 20, 30, 0, -11
        Health:
                HP: 75
        Mobile:
                Speed: 56
        Passenger:
                PipType: Blue
        EngineerRepair:
        RepairsBridges:
        Captures:
                CaptureTypes: building, husk
        -AutoTarget:
        WithInfantryBody:
        RevealsShroud:
                Range: 4c0
        Voiced:
                VoiceSet: AlliedEngineerVoice
```

Or maybe start the line with `#%` or similar to indicate the line is special, or have metainfo such as:
`# last-rule: 20150614`
##

**Comment by Unit158** [comment]
This would also let map devs just pop their map into --upgrade-map without worrying about engine version.

**Comment by Mailaender** [comment]
I wonder why this is required. All traits/files need to be updated to one version for the map/mod to work.

**Comment by abcdefg30** [comment]
It seems like the utility is sometimes adding fields (on traits) another time, if you run the utility before.

**Comment by pchote** [comment]
@abcdefg30 it will only do this if you give it the wrong date.

**Comment by penev92** [comment]
Because the utility can do all sorts of nasty things. My experience is with it ruining maps by upgrading them several times (yeah, my fault for not backing them up). This would be really great for that at least.

**Comment by Phrohdoh** [comment]
@Mailaender, @pchote: take the (often reoccurring) case of RA2.

Either @abcdefg30 or I will run RA2 through the upgrade rules from X to Y and forget to note it in the commit message or will tell the other person but the other person often forgets and thinks to upgrade RA2 in their spare time.

Yes this is entirely our fault but there is no reason this tool can't help imperfect humans.

**Comment by Mailaender** [comment]
Okay. Probably make this optional then with a command line parameter flag so we don't junk up our cleanly maintained repository which is always up to date.

**Comment by pchote** [comment]
The motivations for this have been solved (more or less) with #14964, so closing as fixed.
